README.md
# N27 Smartphone Bank

http://symbolaris.com/info/Facharbeit/Softwareentwicklung.html

Entwicklung einer Smartphone Bank nach dem Vorbild der N26 Bank.

## Was soll die App können?

* Login
* Überweisen
* Kontostand abrufen
* Konto anlegen
* Hilfestellung / Supportdaten / FAQ / Chat
* Kontobewegungen anzeigen
* alle Konten in Übersicht anzeigen
* diverse Einstellungen
* Kredit aufnehmen
* Wertpapierkauf / - verkauf
* Geldanlage
* Kategorien bilden / Ausgaben auswerten
* Daueraufträge erstellen
* Kreditline festlegen

## Welche Dateien, die die Ansicht zeigen sollen, werden benötigt?

Alle Dateien, die etwas anzeigen sollen werden in HTML geschrieben.

* Startseite (index.html)
* login (login.html)
  * Benutzer-ID-Eingabe-Feld
  * Passwort-Eingabe-Feld
  * Bestätigungs-Button


## Realisierung





changelog.md
# Changelog

## 20190506

<input type="text" name="kontonummer" pattern="[0-9]{10}" title="Bitte exakt 10 Ziffern eingeben!">
<input type="number" name="kontonummer" max="10" min="10" required>

## 20190225

* heroku konfiguriert

## 20190218

* Favicon eingebaut
* login ausprogrammiert
* npm-Module installiert


Training/if-else-trainieren.js
/**
 * Hier können Sie IF und ELSE trainieren.
 * Am besten dazu den nodemon starten:
 * 
 *      .\node_modules\.bin\nodemon Training\if-else-training.js
 * 
 * Dann die Ausgabe im Terminal beobachten.
 */

// Deklaration und Initialisierung 
// Lies: "Der Wert 0 wird zugewiesen an
//        eine Variable namens x"
 
let x = 0
let y = 1

if("GW11A".endsWith("A")||"GW11A".endsWith("B")||"GW11A".startsWith("GT")){
    console.log("Die Prüfung ist WAHR. Es werden die Anweisungen im Rumpf von IF ausgeführt.")
}else{
    console.log("Die Prüfung ist FALSCH. Es werden die Anweisungen im Rumpf von ELSE ausgeführt.")
}


// FÄLLE:

// 22. if()

// 21 if ("Gw11a".endsWith("a")) Der Ausdruck ist Wahr. Die Funktion prüft, ob eine Zeichenkette mit einer anderen endet.

// 20. if("GW11A".includes("11")) Der Ausdruck ist Wahr. Die Funktion prüft, ob eine Zeichenkette eine andere Zeichenkette enthält.

// 19. if("GW11A".startsWith("G")) Der Ausdruck ist Wahr. Die Funktion prüft, ob eine Zeichenkette mit einer anderen Zeichenkette beginnt.

// 18. if("GW11A".length === 5) Der Ausdruck ist wahr. Die Eigenschaft length gibt die Anzahl der Zeichen einer Zeichenkette zurück.

// 17. if(undefined)    Der Ausdruck ist falsch. Ein Objekt ist undefined, wenn sie nicht existiert.

// 16. if(1 > 2 || 1 == 1) Der Ausdruck ist wahr, weil eine der beiden Prüfungen wahr ist.
// 15. if(1 > 2 && 1 == 1) Der Ausdruck ist Falsch, weil die erste Prüfung Falsch ist.Alle mit && verketteten Prüfungen müssen Wahr sein, damit die Prüfung insgesamt Wahr wird.
// 14. if(false)      Der Ausdruck ist falsch. 

// 13. if (true)    Der Ausdruck ist wahr.

// 12. if (0)      Die Zahl 0 ist falsch.

// 11. if (-123)   Zahlen sind wahr bis auf 0.

// 10. if("")      leere Zeichenketten in javascript sind falsch.

// 9. if("ah")     nicht leere Zeichenketten in javascript sind wahr.

//8. if(1 != 0)     Vergleichsoperator: "ungleich"       Der Ausdruck ist wahr.

// 7. if(1 <= 1)    Vergleichsoperator: "kleiner gleich"    Der Ausdruck ist wahr.

// 6. if(1 >= 1)    Vergleichsoperator: "größer gleich"   Der Ausdruck ist wahr.

// 5. if(1 === "1") Vergleichsoperator: "gleich"          Der Ausdruck ist falsch.
//                  Das 3fache "===" prüft auf Gleichheit des Wertes und des Typs.
//                  "1" : String / Zeichenkette
//                   1  : Zahl 

// 4. if(1 == "1")  Vergleichsoperator: "gleich"          Der Ausdruck ist wahr.
//                  Das doppelte "==" prüft auf Gleichheit des Wertes,
//                  aber nicht auf Gleichheit des Typs.

// 3. if(0 == 1)    Vergleichsoperator: "gleich"          Der Ausdruck ist falsch.

// 2. if(0 > 1)      Vergleichsoperator: "größer als"     Der Ausdruck ist falsch.

// 1. if(0 < 1)      Vergleichsoperator: "kleiner als"    Der Ausdruck ist wahr.

Training/klasse-objekt-ejs-trainieren.js
const express = require('express')
const bodyParser = require('body-parser')
const app = express()
app.set('view engine', 'ejs')
app.use(bodyParser.urlencoded({extended: true}))
app.set('views', 'Training')

const server = app.listen(process.env.PORT || 3000, () => {
    console.log('Server lauscht auf Port %s', server.address().port)    
})

// Eine Klasse ist ein Bauplan. Der Bauplan sieht vor, wie Objekte erstellt werden.
// Alle Objekte, die von einem Bauplan erstellt werden, haben die selben Eigenschaften,
// aber möglicherweise unterschiedliche Eigenschaftswerte.

// Klassendefintion
// ================

class Rechteck{
    constructor(){
        this.laenge
        this.breite
    }
}

// Klassendefinition für Schüler in einer Schule:

class Schueler{
    constructor(){
        this.geschlecht
        this.klasse
        this.alter
        this.vorname
        this.nachname
    }
}

class Fussballer{
    constructor(){
        this.vorname 
        this.nachname
        this.mannschaft
        this.geschlecht
    }
}

// Deklaration eines neuen Objekts vom Typ Rechteck
// Deklaration = Bekanntmachung
// let rechteck = ...

// Instanziierung eines neuen Objekts.
// Instanziierung erkennt man immer am reservierten Wort "new".
// Bei der Instanziierung wird Arbeitsspeicher bereitgestellt.
// ... = new Rechteck()

// 1.Deklaration  2.Instanziierung
let rechteck = new Rechteck()
let schueler = new Schueler()
let fussballer = new Fussballer()


// 3. Initialisierung (Konkrete Eingeschaftswerte werden zugewiesen)

rechteck.breite = 2
rechteck.laenge = 3

schueler.geschlecht = "w"
schueler.alter = 17

fussballer.mannschaft = "FC Borken"
fussballer.vorname = "Firke"

console.log("Länge: " + rechteck.laenge)
console.log("Breite: " + rechteck.breite)


// Wenn localhost:3000/klasse-objekt-ejs-trainieren aufgerufen wird ...

app.get('/klasse-objekt-ejs-trainieren',(req, res, next) => {   

    // ... wird klasse-objekt-ejs-trainieren.ejs gerendert:

    res.render('klasse-objekt-ejs-trainieren', {                                      
        breite : rechteck.breite,
        laenge : rechteck.laenge,
        geschlecht : schueler.geschlecht,
        alter : schueler.alter,
        vorname : fussballer.vorname,
        mannschaft : fussballer.mannschaft
    })
})


server.js
class Konto{
    constructor(){
        this.Kontonummer
        this.Kontoart
    }
}

const express = require('express')
const bodyParser = require('body-parser')
const cookieParser = require('cookie-parser')
const app = express()
app.set('view engine', 'ejs')
app.use(express.static('public'))
app.use(bodyParser.urlencoded({extended: true}))
app.use(cookieParser())

const server = app.listen(process.env.PORT || 3000, () => {
    console.log('Server lauscht auf Port %s', server.address().port)    
})

app.get('/',(req, res, next) => {   

    let idKunde = req.cookies['istAngemeldetAls']
    
    if(idKunde){
        console.log("Kunde ist angemeldet als " + idKunde)
        res.render('index.ejs', {                              
        })
    }else{
        res.render('login.ejs', {                    
        })    
    }
})

// Wenn die Seite localhost:3000/impressum aufgerufen wird, ...

app.get('/impressum',(req, res, next) => {   

    let idKunde = req.cookies['istAngemeldetAls']
    
    if(idKunde){
        console.log("Kunde ist angemeldet als " + idKunde)
        
        // ... dann wird impressum.ejs gerendert.
        
        res.render('impressum.ejs', {                              
        })
    }else{
        res.render('login.ejs', {                    
        })    
    }
})

app.get('/login',(req, res, next) => {         
    res.cookie('istAngemeldetAls', '')       
    res.render('login.ejs', {                    
    })
})

app.post('/',(req, res, next) => {   
    
    const idKunde = req.body.idKunde
    const kennwort = req.body.kennwort
        
    if(idKunde === "4711" && kennwort === "123"){            
        console.log("Der Cookie wird gesetzt:")
        res.cookie('istAngemeldetAls', idKunde)
        res.render('index.ejs', {  
            kunde : idKunde          
        })
    }else{            
        console.log("Der Cookie wird gelöscht")
        res.cookie('istAngemeldetAls','')
        res.render('login.ejs', {                    
        })
    }
})

// Wenn die Seite localhost:3000/kontoAnlegen angesurft wird, ...

app.get('/kontoAnlegen',(req, res, next) => {   

    let idKunde = req.cookies['istAngemeldetAls']
    
    if(idKunde){
        console.log("Kunde ist angemeldet als " + idKunde)
        
        // ... dann wird kontoAnlegen.ejs gerendert.
        
        res.render('kontoAnlegen.ejs', {    
            meldung : ""                          
        })
    }else{
        res.render('login.ejs', {                    
        })    
    }
})

// Wenn der Button auf der kontoAnlegen-Seite gedrückt wird, ...

app.post('/kontoAnlegen',(req, res, next) => {   

    let idKunde = req.cookies['istAngemeldetAls']
    
    if(idKunde){
        console.log("Kunde ist angemeldet als " + idKunde)
        
        let konto = new Konto()
        konto.Kontonummer = req.body.kontonummer
        konto.Kontoart = req.body.kontoart
              
        res.render('kontoAnlegen.ejs', {                              
            meldung : "Das " + konto.Kontoart + " mit der Kontonummer " + konto.Kontonummer + " wurde erfolgreich angelegt."
        })
    }else{
        res.render('login.ejs', {                    
        })    
    }
})